core/storage/page_bitmap.rs

```diff
@@ -0,0 +1,279 @@
+                break;
+            }
+        }
+        None
+    }
+
+    /// Search for a free run of `need` pages beginning from `start`
+    fn find_free_run(&self, start: u32, need: u32) -> Option<u32> {
+        let mut pos = start;
+        let limit = self.n_pages.saturating_sub(need - 1);
+
+        while pos < limit {
+            if let Some(next_free) = self.next_free_bit_from(pos) {
+                if next_free + need > self.n_pages {
+                    break;
+                }
+                if self.check_run_free(next_free, need) {
+                    return Some(next_free);
+                }
+                pos = next_free + 1;
+            } else {
+                break;
+            }
+        }
+        None
+    }
+
+    /// Frees a contiguous run of pages, marking them as available for reallocation
+    /// and update the scan hints to potentially reuse the freed space in future allocations.
+    pub fn free_run(&mut self, start: u32, count: u32) {
+        if count == 0 {
+            return;
+        }
+        self.mark_run(start, count, true);
+        // Update scan hint to potentially reuse this space
+        if start < self.scan_run_low {
+            self.scan_run_low = start;
+        } else if start > self.scan_one_high {
+            // If we freed a run beyond the current scan hint, adjust it
+            self.scan_one_high = start.saturating_add(count).saturating_sub(1);
+        }
+    }
+
+    /// Checks whether a contiguous run of pages is completely free.
+    pub fn check_run_free(&self, start: u32, len: u32) -> bool {
+        if start.saturating_add(len) > self.n_pages {
+            return false;
+        }
+        self.inspect_run(start, len, |word, mask| (word & mask) == mask)
+    }
+
+    /// Marks a contiguous run of pages as either free or allocated.
+    pub fn mark_run(&mut self, start: u32, len: u32, free: bool) {
+        let (mut word_idx, bit_offset) = Self::page_to_word_and_bit(start);
+        let mut remaining = len as usize;
+        let mut pos_in_word = bit_offset as usize;
+
+        while remaining > 0 {
+            let bits_in_word = (Self::WORD_BITS as usize).saturating_sub(pos_in_word);
+            let bits_to_process = remaining.min(bits_in_word);
+
+            let mask = if bits_to_process == Self::WORD_BITS as usize {
+                Self::ALL_FREE
+            } else {
+                (1u64 << bits_to_process).saturating_sub(1) << pos_in_word
+            };
+
+            if free {
+                self.words[word_idx] |= mask;
+            } else {
+                self.words[word_idx] &= !mask;
+            }
+
+            remaining -= bits_to_process;
+            // move to the next word/reset position
+            word_idx += 1;
+            pos_in_word = 0;
+        }
+    }
+
+    /// Process a run of bits with a read-only operation
+    fn inspect_run<F>(&self, start: u32, len: u32, mut check: F) -> bool
```

Comment:
Can we instead do the logic of `check()` inline inside `inspect_run`, since it's only called from one place, and the closure it is currently called with captures nothing from the environment?